Add labels for some error messages

scoopr opened the pull request:
This places some pieces on how to transform the ids in the errors to a label, and adds them to the validation message.

This changes the formatting of the validation error to happen when constructing the error, rather than in the unhandled error handler.
It also requires some code for all the error variants for extracting the ids.

scoopr commented:
As discussed in gfx-rs/wgpu#931 and on matrix, this tries to do the label formatting on the wgpu-rs side. Depends on the changes on #931.

I'm not entirely happy with it, but it certainly is workable.

kvark commented on src/backend/error.rs:
```diff
@@ -0,0 +1,83 @@
+    }
+}
+
+fn format_error_line(err: &dyn Display) -> String {
```
these should probably work with `fmt::Arguments` parameter instead of `&dyn Display`, see https://doc.rust-lang.org/std/fmt/struct.Arguments.html

kvark commented on src/backend/direct.rs:
```diff
@@ -16,9 +16,15 @@ use std::{
 };
 use typed_arena::Arena;
 
+use crate::backend::error::format_error;
```
nit: can we merge this include with the other `use crate::`?

kvark reviewed (commented):
Thank you!
It looks to be a fine way to proceed, just a few notes to address :+1:

scoopr commented:
The latest commit now implements the labels for the errors I found that had Ids to resolve.
Also formats the labels in the same format with the `ContextError`. Also had to make ContextError box the source error, as I needed a concrete type to downcast to.
This requires gfx-rs/wgpu#1033 to land first

With this, a error could look like this
```
wgpu error: Validation Error

Caused by:
    In CommandEncoder::copy_buffer_to_buffer
    Copy error
    destination buffer/texture is missing the `COPY_DST` usage flag
      note: destination = `staging_buffer`
```

The same in release mode looks like
```
wgpu error: Validation Error

Caused by:
    In CommandEncoder::copy_buffer_to_buffer
    Copy error
    destination buffer/texture is missing the `COPY_DST` usage flag
      note: destination = `<Buffer-(0, 1, Metal)>`
```

The creation/ContextError looks like
```
wgpu error: Validation Error

Caused by:
    In Device::create_render_pipeline
      note: label = `hello-triangle pipeline`
    error in stage VERTEX
    unable to find an entry point at VERTEX stage
```

scoopr commented:
Oh it seems I neglected few points above that might still be relevant

scoopr commented:
Addressed most of the comments above, except for the `fmt::Arguments` thing. I see the value in it, but I don't have the stamina to work out the details of it right now, it seems to need some larger restructuring to be useful here.

scoopr commented on src/backend/error.rs:
```diff
@@ -0,0 +1,83 @@
+
+use super::Context;
+
+pub(crate) fn format_error(err: &(dyn Error + 'static), context: &Context) -> String {
```
I'm fine either way, so I changed it to `impl`. I guess I originally thought that this code doesn't need to be monomorphised, and I figured it should emit only single body with the `dyn`. Though I guess with more vtable handling.

kvark commented on src/backend/error.rs:
```diff
@@ -0,0 +1,242 @@
+}
+
+fn fmt_pretty_any(error: &(dyn Error + 'static), context: &Context) -> String {
+    if let Some(pretty_err) = error.downcast_ref::<super::direct::ContextError>() {
```
can this be casted to a `PrettyError` trait object?

scoopr replied:
Needs to be a concrete type
> the `downcast_ref` method cannot be invoked on a trait object

kvark commented on src/backend/error.rs:
```diff
@@ -0,0 +1,242 @@
+                let name = wgc::gfx_select!(id => global.buffer_label(id));
+                ret.push_str(&format_label_line("buffer", &name));
+            }
+            Self::DestroyedBuffer(id) => {
```
should probably merge this with `Self::Buffer` path?

kvark reviewed (commented):
Thank you! We are almost there!
I'll be fine following up with the FormatArgs. In fact, this code may need to be redone after https://github.com/gfx-rs/wgpu/issues/1032 is done, anyway.

kvark commented:
> This requires gfx-rs/wgpu#1033 to land first

we can wait for it to land in a few minutes :)

scoopr commented:
update `wgpu-core` rev, merged the match arm.

kvark commented:
Great, thank you!
bors r+
